modules/core-components/src/test/java/org/mule/runtime/core/privileged/processor/chain/DefaultMessageProcessorChainTestCase.java

```diff
@@ -0,0 +1,304 @@
+
+    flow = builder("flow", muleContext).processingStrategyFactory(processingStrategyFactory).build();
+    flow.setAnnotations(singletonMap(LOCATION_KEY, from("flow")));
+    flow.initialise();
+    flow.start();
+  }
+
+  @Override
+  protected Map<String, Object> getStartUpRegistryObjects() {
+    return singletonMap(REGISTRY_KEY, componentLocator);
+  }
+
+  @After
+  public void after() throws MuleException {
+    flow.stop();
+    flow.dispose();
+  }
+
+  @Test
+  public void all() throws Exception {
+    createAndRegisterFlow(muleContext, APPLE_FLOW, componentLocator);
+    ScatterGatherRouter scatterGatherRouter = new ScatterGatherRouter();
+    scatterGatherRouter.setAnnotations(getAppleFlowComponentLocationAnnotations());
+    scatterGatherRouter
+        .setRoutes(asList(newChain(empty(), getAppendingMP("1")), newChain(empty(), getAppendingMP("2")),
+                          newChain(empty(), getAppendingMP("3"))));
+
+    CoreEvent event = getTestEventUsingFlow("0");
+    final MessageProcessorChain chain = newChain(empty(), singletonList(scatterGatherRouter));
+    Message result = process(chain, CoreEvent.builder(event).message(event.getMessage()).build()).getMessage();
+    assertThat(result.getPayload().getValue(), instanceOf(Map.class));
+    Map<String, Message> resultMessage = (Map<String, Message>) result.getPayload().getValue();
+    assertThat(resultMessage.values().stream().map(msg -> msg.getPayload().getValue()).collect(toList()).toArray(),
+               is(equalTo(new String[] {"01", "02", "03"})));
+
+    scatterGatherRouter.stop();
+    scatterGatherRouter.dispose();
+  }
+
+  @Test
+  public void choice() throws Exception {
+    ChoiceRouter choiceRouter = new ChoiceRouter();
+    choiceRouter.setAnnotations(getAppleFlowComponentLocationAnnotations());
+    choiceRouter.setAnnotations(singletonMap(LOCATION_KEY, TEST_CONNECTOR_LOCATION));
+    choiceRouter.addRoute("true", newChain(empty(), getAppendingMP("1")));
+    choiceRouter.addRoute("true", newChain(empty(), getAppendingMP("2")));
+    choiceRouter.addRoute("true", newChain(empty(), getAppendingMP("3")));
+    initialiseIfNeeded(choiceRouter, muleContext);
+
+    try {
+      assertThat(process(newChain(empty(), choiceRouter), getTestEventUsingFlow("0")).getMessage().getPayload().getValue(),
+                 equalTo("01"));
+    } finally {
+      disposeIfNeeded(choiceRouter, LOGGER);
+    }
+  }
+
+  @Override
+  protected CoreEvent process(Processor messageProcessor, CoreEvent event) throws Exception {
+    initialiseIfNeeded(messageProcessor, muleContext);
+    startIfNeeded(messageProcessor);
+
+    try {
+      return super.process(messageProcessor, event);
+    } finally {
+      final SchedulerService schedulerService = muleContext.getSchedulerService();
+    }
+  }
+
+  private AppendingMP getAppendingMP(String append) {
+    return new NonBlockingAppendingMP(append);
+  }
+
+  class NonBlockingAppendingMP extends AppendingMP {
```

Comment:
make inner class static

Reply:
cannot do, it uses `nonBlockingProcessorsExecuted` from the outer class